File: java/mergeSort.java
import java.io.*;
import java.util.Scanner;

public class mergeSort {
    public void merge_sort(int[] array, int left, int right) {
        if (right <= left) return;
        int mid = (left+right)/2;
        merge_sort(array, left, mid);
        merge_sort(array, mid+1, right);
        merge(array, left, mid, right);
    }

    void merge(int[] array, int left, int mid, int right) {
        int lengthLeft = mid - left + 1;
        int lengthRight = right - mid;

        int leftArray[] = new int [lengthLeft];
        int rightArray[] = new int [lengthRight];

        for (int i = 0; i < lengthLeft; i++)
            leftArray[i] = array[left+i];
        for (int i = 0; i < lengthRight; i++)
            rightArray[i] = array[mid+i+1];

        int leftIndex = 0;
        int rightIndex = 0;

        for (int i = left; i < right + 1; i++) {
            if (leftIndex < lengthLeft && rightIndex < lengthRight) {
                if (leftArray[leftIndex] < rightArray[rightIndex]) {
                    array[i] = leftArray[leftIndex];
                    leftIndex++;
                }
                else {
                    array[i] = rightArray[rightIndex];
                    rightIndex++;
                }
            }
            else if (leftIndex < lengthLeft) {
                array[i] = leftArray[leftIndex];
                leftIndex++;
            }
            else if (rightIndex < lengthRight) {
                array[i] = rightArray[rightIndex];
                rightIndex++;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        final int N = 1000;
        File array = new File("array.txt");
        Scanner scanner = new Scanner(array);
        int [] arr = new int [N];
        int i = 0;

        while(scanner.hasNextInt()){
            arr[i++] = scanner.nextInt();
        }
        mergeSort obj = new mergeSort();
        final double startTime = System.nanoTime() / 1000000;
        obj.merge_sort(arr, 0, arr.length - 1);
        final double endTime = System.nanoTime() / 1000000;

        System.out.println((endTime - startTime));
    }
}

File: kt/mergeSort.kt
import java.io.File
import java.util.*


fun merge(arr: IntArray, l: Int, m: Int, r: Int): Unit{
    val n1 = m - l + 1
    val n2 = r - m
    val L = IntArray(n1)
    val R = IntArray(n2)
    for (i in 0 until n1){
        L[i] = arr[i + l]
    }
    for (j in 0 until n2){
        R[j] = arr[j + m + 1];
    }
    var i = 0
    var j = 0
    var k = l
    while (i < n1 && j < n2){
        if (L[i] <= R[j]){
            arr[k] = L[i]
            i++
        }
        else{
            arr[k] = R[j]
            j++
        }
        k++
    }
    while(i < n1){
        arr[k] = L[i]
        i++
        k++
    }
    while(j < n2){
        arr[k] = R[j]
        j++
        k++
    }
}


fun merge_sort(arr: IntArray, l: Int, r: Int){
    if (l < r){
        var m: Int = l + (r - l)/2
        merge_sort(arr, l, m)
        merge_sort(arr, m + 1, r)
        merge(arr, l, m, r)
    }
}

fun main(args : Array<String>){

    val N = 10000
    val arrayContent = File("./array.txt")
    var scanner = Scanner(arrayContent)
    val arr = IntArray(N)
    var i = 0

    while (scanner.hasNextInt()) {
        arr[i++] = scanner.nextInt()
    }

    val time = System.nanoTime();
    merge_sort(arr, 0, arr.lastIndex)

    val timeMilli = (System.nanoTime() - time) / 1000000.0f

    println("$timeMilli")

}
